builder/api/README.md

```diff
@@ -0,0 +1,173 @@
+     be `List<SecurityProvider>`, where `SecurityProvider extends NamedService` and  
+     `SecurityProviderProvider extends ConfiguredProvider<SecurityProvider>`
+
+This will expect the following configuration:
+```yaml
+security-providers:
+  discover-services: true # optional, to override "providerDiscoverServices" option
+  providers:
+    - name: "my-provider"
+      type: "http-basic"
+      enabled: true
+```
+
+The generated code will read all nodes under `providers` and map them to an instance.  
+
+## Config driven
+
+Only with Pico!!!
+
+We can use configuration to drive discovery of "beans". For such cases, we need two things:
+1. The prototype must be marked as `@io.helidon.pico.configdriven.ConfigBean`
+2. The implementation of a contract must be marked as `@io.helidon.prico.configdrivent.ConfigDriven(ConfigBean.class)`
+
+Now the config bean will be instantiated based on the rules of the `@Configured` and `@ConfigBean` annotations, basic rules:
+- If the configuration node (root=true, prefix=`node`) exists in configuration, a config bean will be created for it
+- If (wantDefault=`true`), a config bean will be created based on `@default` name (if present in configuration), or with all defaults, this bean will be available with qualifier `@Default`, also `@Named("@default")` (both are not required)
+- If (repeatable=`true`), a config bean will be crated for each child node of the `prefix` node, with qualifier `@Named`, where `name` is either node child `name` as a string, or the name of the config node itself
+
+And config driven bean (something that is `@ConfigDriven(SomeConfigBean.class)`) will be created for each config bean of appropriate type with matching qualifiers (esp. `@Default` and `@Named`). When an injection point exists on such a config driven bean that is of the config bean type, it will be satisfied by the config bean driving this instance 
+
+A config driven type is a type that is generated from `@Prototype.Blueprint` and is `@Configured` with `root=true`.
+These types can be discovered automatically when Pico is used.
+
+The metadata needed:
+
+- config key (such as `fault-tolerance.retries`, `server`)
+- if repeatable (`true` for `RetryPrototype`, `false` for `server`)
+
+How do we discover such types? 
+
+```java
+// wantDefault - create default instance EVEN if configured instances exist
+// atLeastOne - either from configuration, or use default (if any required fields exist, this would fail for default)
+@ConfigDriven(value = LoomServer.class, activateByDefault = true, atLeastOne = true, repeatable = false, wantDefault = false)
+@ConfiguredBy(ServerConfig.class)
+@Contract
+public interface WebServer extends RuntimeType<ServerConfig> {
+} 
+```
+
+This would require code somewhere that would crate the instances from config:
+
+1. Builder must implement `String configKey()` method
+
+```java
+// if "at least one"
```

Comment:
i'm not sure we should start with "at least one". That is more of a javadoc thing on validation. I think if we are going to say anything, it is better to have `wantDefault` covered.

Reply:
pushed